src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -808,8 +808,19 @@ var ReactCompositeComponentMixin = {
   attachRef: function(ref, component) {
     var inst = this.getPublicInstance();
     invariant(inst != null, 'Stateless function components cannot have refs.');
+    var publicComponentInstance = component.getPublicInstance();
+    var componentName = component && component.getName ?
+      component.getName() : 'a component';
+    warning(publicComponentInstance != null,
+      'Stateless function components cannot be given refs ' +
+      '(See ref "%s" in %s created by %s). ' +
+      'Attempts to access this ref will fail.',
+      ref,
+      componentName,
+      this.getName()
+    );
```

Comment:
This whole warning should be within a `__DEV__` block so that it doesn't need to be evaluated in the production builds.